### ISSUE TITLE
Document Docker Experience

### ISSUE THREAD

**Issue by richlander** [dotnet/core-docs#806]
We need to document more of the E2E docker experience.

We have this great doc already, but need more: https://github.com/dotnet/core-docs/blob/master/docs/csharp/tutorials/microservices.md

Here's what we need:

- Most basic "hello world" with docker and .NET.
- Most optimal "hello world" with docker and .NET. Assumption is that this will be a standalone app. This should be easy to switch to Alpine when we support it.
- More E2E tutotials. The linked example above is a great one.
- The tutorials should cover Linux and Windows scenarios.
- Conceptual docs that describes our docker story and the concepts you need to know. This should presumable live at ~/core/docker. 
- Update our "dotnet docker" page to ensure that it is a great starting point and links to this great documentation, presumably to ~/core/docker. See: https://github.com/dotnet/dotnet-docker. See: https://hub.docker.com/r/microsoft/dotnet/

/cc @dleeapho, @BillWagner

**Comment by BillWagner** [comment]
Let's discuss how to cover this best across different platforms that developers might choose. From the reading I've done so far, developers could create .NET apps targeting Docker using a Linux workstation, a Mac, or a Windows PC. 

Also, I've seen blogs indicating that windows based VMs for Docker are also coming.

That's a large matrix to cover. We don't want to duplicate content and confuse readers. And we don't want to imply by their absence that supported scenarios don't work. 

I'll start on an outline of ideas.

**Comment by dleeapho** [comment]
Sounds good. I think we also want to talk about how to use the -sdk, -core, and -core-deps images. @blackdwarf and @glennc had some content they were collaborating on on that subject.  
/cc @michaelsimons

**Comment by BillWagner** [comment]
That's great. I had a couple conversations with @michaelsimons earlier in the spring regarding both of those images. Those are definitely scenarios we want to cover. 

I'll cc folks on the outline. I expect I'll have that in place early next week.

**Comment by spboyer** [comment]
@dleeapho the doc : https://docs.microsoft.com/en-us/dotnet/articles/core/docker/building-net-docker-images has just been published to cover those scenarios.

**Comment by dleeapho** [comment]
Thanks @spboyer ! @michaelsimons we should link to the above article from the microsoft/dotnet readme.md

**Comment by mairaw** [comment]
@spboyer @richlander what else is left to do here? Should this become a project?

**Comment by spboyer** [comment]
@mairaw Yes this needs to be a project. I have an outline of topics. @kendrahavens is also working on the samples repo.  

@richlander @SteveLasker

**Comment by BillWagner** [comment]
We have some history and ideas here. Let's continue to use this issue as the tracking issue for the Docker E2E project. Add links between this issue and the project when that is created.

**Comment by spboyer** [comment]
Project created https://github.com/dotnet/docs/projects/4
@BillWagner @mairaw @richlander

**Comment by spboyer** [comment]
Outline of topics that has been circulating

- [ ] Hello World – Basic – File -> New Project using VS2017 Docker Tools for VS
- [ ] Using Docker for Console Applications (WebJobs)
- [ ] Web Application using multiple containers
- [ ] Scaling Using Compose
- [ ] Building Applications using Microservices
         * Web Application
         * Web API Container
         * Database
         * Redis / Cache
- [ ] CI / CD Fundamentals
- [ ] VSTS
- [ ] Testing
- [ ] Docker on Windows
         * .NET Core on Nano Server
         * .NET 4.6 Applications
- [ ] Docker on Azure
- [ ] Deployment / Publishing
- [ ] Debugging

**Comment by kendrahavens** [comment]
This list looks good.

These could use the samples we already have on docker hub and gh below: 
- [ ] Docker on Windows
  * [.NET Core on Nano Server](https://github.com/dotnet/dotnet-docker-samples)
  * [.NET 4.6 Applications](https://github.com/Microsoft/dotnet-framework-docker-samples/tree/master/dotnetapp-4.6.2)

For these, I'm working on something with the SQL Server Team that overlaps this topic. I'll share it when its further along next week.
- [ ] Building Applications using Microservices
  * Web Application
  * Web API Container
  * Database
  * Redis / Cache

**Comment by spboyer** [comment]
@kendrahavens any progress on the item you mentioned for Microservices?

**Comment by kendrahavens** [comment]
@LuisBosquez is working the [Database scenario](https://github.com/LuisBosquez/docker-mssql-samples)
I've worked on a very simple [Web Application for ASP.NET Core](https://github.com/kendrahavens/dockerize-dotnet-app).

The Web API Container and Redis/Cache topics would still be up for grabs.

**Comment by squillace** [comment]
Hey, all, from the linux/mac side of the house, anything that can streamline the CI/CD scenario with non-M$ tools/services is great stuff. pain points ARE definitely those that are hard to "get". I'd be onboard with making sure it all works from linux/mac and _through_ any kubernetes/dcos/swarm scenario, especially on Azure (but not explicitly). I've written for dcos.io and docker and soon will be for kubernetes.....

**Comment by mairaw** [comment]
Let's turn this into a project and come up with a list of topics/work items that need to happen to make progress on this.

**Comment by squillace** [comment]
Happy to help. we already have dotnet core docs for csharp and fsharp with core on Linux containers out there..... My team does the kubernetes code extension, and share work on the docker extension for core.

**Comment by SteveLasker** [comment]
Happy to help.
ACR Build: https://aka.ms/acr/build supports simple, local scoped/pre-commit builds to git-commit, automated builds with OS & Framework patching. We just added docker hub based images, with ARM qemu builds for IoT scenarios.  We'll have windows container builds, with all supported bands, 1709, 1803, ... coming later this month.

**Comment by JRAlexander** [comment]
This item is the discussion item for the original project: https://github.com/dotnet/docs/projects/4#card-4672260, and the original items assigned were completed.  There was talk around 1Q18 about adding more but that was pushed off and I didn't close it at that time.  We can add new items to it, but I'd rather close this one and start a new docker items project fresh.
